keps/sig-network/2009-dns-autopath/README.md

```diff
@@ -0,0 +1,228 @@
+
+These search paths are set to make sure: 
+
+ 1. Pods can discover Services in the same namespace using just the service name.
+ 2. Pods can discover Services across namespaces using shorthand of the form "$SVCNAME.$NSNAME"
+ 3. Pods can discover resources within the same cluster.
+
+These searchpaths are included in pods' /etc/resolv.conf by kubelet and are enforced by setting ndots to 5. This means any hostname lookups with fewer than 5 dots will be expanded using all the search paths listed.
+
+When pod issues a query to lookup hostname "service123", it is expanded to 6 queries - one for the original hostname and one with each of the searchpaths appended. Some resolvers issue both A and AAAA queries, so this can be a total of 12 or more queries for every single DNS lookup. When these queries are issued in parallel, they end up at the node with the same source tuple and need to be DNAT'ed increasing the chance of a [netfilter race condition](https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts).
+If a response is not received for one of the queries, the DNS lookup on the client side will fail after a 5s timeout. This issue is most-applicable to A/AAAA lookups which form the bulk of DNS lookups. PTR Records are not subject to searchpath expansion(since they always have >5 dots).
+
+### Goals
+
+* Provide a solution to minimize number of DNS queries on the client side, for a DNS lookup. 
+This solution will preserve the current behavior that allows shortname lookups of Kubernetes resources.
+
+* Make this solution configurable via the k8s API.
+
+### Non-Goals
+Modifying the current Kubernetes DNS Schema is not a goal of this KEP.
+
+
+## Proposal
+
+This proposal introduces use of a single searchpath by client pods when performing a DNS lookup. This searchpath will contain all the information needed to figure out the list of searchpaths to apply for that DNS lookup. 
+
+As observed from the list of searchpaths in the current DNS schema, the searchpaths used for each pod is the same, except for the namespace. If this information can be included in a single searchpath, some entity that is aware of the DNS schema can expand this to a list of searchpaths. This entity could be the ClusterDNS server(CoreDNS), NodeLocal DNSCache or a sidecar container running in the client pod.
+ 
+The new searchpath is of the form: 
+`search.$NS.$SUFFIX.ap.k8s.io`, where $NS is the namespace of the pod and $SUFFIX is the cluster suffix. ap.k8s.io is the delimiter to identify if a query needs search expansion.
+
+This searchpath can be set by using `dnsPolicy:None` to start with, as described [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config).
+Once this feature graduates to GA, the default searchpath for `ClusterFirst` and `ClusterFirstWithHostNet` will include this new searchpath instead of the 3 different k8s specific searchpaths.
```

Comment:
I am not sure how we can do that - we need to know that the cluster's DNS resolver supports it (either query parsing or or EDNS0) so it seems like it has to be opt-in basically forever.  It might be possible to change the default behavior on a cluster-by-cluster basis but not globally.